predicators/envs/narrow_passage.py

```diff
@@ -0,0 +1,490 @@
+        door = objects[0]
+        return state.get(door, "open") == 1
+
+    def _coords_out_of_bounds(self, new_x: float, new_y: float) -> bool:
+        if (self.x_lb <= new_x <= self.x_ub) and (self.y_lb <= new_y <=
+                                                  self.y_ub):
+            return False
+        return True
+
+    def _state_has_collision(self, state: State) -> bool:
+        robot, = state.get_objects(self._robot_type)
+        robot_geom = self._object_to_geom(robot, state)
+        # Check for collisions with obstacles.
+        for obstacle in state.get_objects(self._wall_type):
+            obstacle_geom = self._object_to_geom(obstacle, state)
+            if robot_geom.intersects(obstacle_geom):
+                return True
+        # Check for collisions with closed doors.
+        door, = state.get_objects(self._door_type)
+        if not self._DoorIsOpen_holds(state, [door]):
+            door_geom = self._object_to_geom(door, state)
+            if robot_geom.intersects(door_geom):
+                return True
+        return False
+
+    def _robot_near_door(self, state: State) -> bool:
+        """Returns true if the robot is within range of the door sensor."""
+        robot, = state.get_objects(self._robot_type)
+        robot_geom = self._object_to_geom(robot, state)
+        door_sensor, = state.get_objects(self._door_sensor_type)
+        door_sensor_geom = self._object_to_geom(door_sensor, state)
+        return robot_geom.intersects(door_sensor_geom)
+
+    def _object_to_geom(self, obj: Object, state: State) -> _Geom2D:
+        """From doors.py."""
+        x = state.get(obj, "x")
+        if (obj.is_instance(self._robot_type)
+                or obj.is_instance(self._target_type)):
+            y = state.get(obj, "y")
+            return utils.Circle(x, y, self.robot_radius)
+        # Cache static objects such as door and walls
+        if obj not in self._static_geom_cache:
+            if obj.is_instance(self._door_sensor_type):
+                y = self.y_lb + (self.y_ub - self.y_lb) / 2
+                self._static_geom_cache[obj] = utils.Circle(
+                    x, y, self.door_sensor_radius)
+            else:
+                if obj.is_instance(self._wall_type):
+                    y = self.y_lb + (self.y_ub -
+                                     self.y_lb) / 2 - self.wall_thickness_half
+                    width = state.get(obj, "width")
+                    height = self.wall_thickness_half * 2
+                else:
+                    assert obj.is_instance(self._door_type)
+                    y = self.y_lb + (
+                        self.y_ub - self.y_lb
+                    ) / 2 - self.wall_thickness_half + self.doorway_depth
+                    width = (self.robot_radius + self.door_width_padding) * 2
+                    height = (self.wall_thickness_half -
+                              self.doorway_depth) * 2
+                self._static_geom_cache[obj] = Rectangle(x=x,
```

Comment:
you can do `utils.Rectangle` and avoid the import, like we're doing with `utils.Circle` above